ssz_serialization/merkleization.nim

```diff
@@ -746,3 +1169,217 @@ func hash_tree_root*(x: auto): Digest =
       hashTreeRootAux(toSszType(x))
 
   trs "HASH TREE ROOT FOR ", name(typeof(x)), " = ", "0x", $result
+
+# Note: If this function is also called `hash_tree_root`, the Nim compiler may
+# produce incorrect code when calling it, e.g., when called by this overload:
+#   func hash_tree_root*(
+#       x: auto,
+#       indices: static openArray[GeneralizedIndex],
+#       roots: var openArray[Digest]): Result[void, string]
+#
+# Instead of passing the static `indices` to this function, the Nim compiler
+# generates a copy of `indices` with incorrect types (pointers instead of NU64).
+# This copy is then passed to this function which then interprets it as NU64[].
+# On 32-bit platforms such as i386 the pointer width does not match NU64 width,
+# leading to incorrect followup computations and out-of-bounds memory access.
+#
+# Creating a minimal reproducible example showcasing this bug proved difficult.
+#
+# Affected: Nim Compiler Version 1.2.14 [Linux: i386]
+# https://github.com/nim-lang/Nim/issues/19157
+func hash_tree_root_multi(
+    x: auto,
+    indices: openArray[GeneralizedIndex],
+    roots: var openArray[Digest],
+    loopOrder: seq[int],
+    slice: Slice[int],
+    atLayer = 0): Result[void, string] =
+  trs "STARTING HASH TREE ROOT FOR TYPE ", name(typeof(x)),
+    slice.mapIt(indexAt(it))
+  mixin toSszType
+
+  when x is HashArray|HashList:
+    ? hashTreeRootCached(x, indices, roots, loopOrder, slice, atLayer)
+  else:
+    ? hashTreeRootAux(toSszType(x), indices, roots, loopOrder, slice, atLayer)
+
+  trs "HASH TREE ROOT FOR ", name(typeof(x)),
+    slice.mapIt(indexAt(it)), " = ", slice.mapIt("0x" & $rootAt(it))
+  ok()
+
+template normalize(v: GeneralizedIndex): GeneralizedIndex =
```

Comment:
Double evaluation of `v`

Reply:
This template is only used by the `proc` immediately following it, and it is not passing computed expressions to the template. So, the double evaluation should be alright.

```
template normalize(v: GeneralizedIndex): GeneralizedIndex =
  ((v shl 1) or 1) shl leadingZeros(v)

proc cmpDepthFirst(x, y: GeneralizedIndex): int =
  cmp(x.normalize, y.normalize)
```